src/dodal/devices/i04/transfocator.py

```diff
@@ -18,65 +22,75 @@ class Transfocator(Device):
         my_transfocator.set(vert_beamsize_microns)
     """
 
-    beamsize_set_microns = Cpt(EpicsSignal, "VERT_REQ", kind=Kind.hinted)
-    predicted_vertical_num_lenses = Cpt(EpicsSignal, "LENS_PRED")
-
-    number_filters_sp = Cpt(EpicsSignal, "NUM_FILTERS")
-
-    start = Cpt(EpicsSignal, "START.PROC")
-    start_rbv = Cpt(EpicsSignalRO, "START_RBV")
-
-    vertical_lens_rbv = Cpt(EpicsSignalRO, "VER", kind=Kind.hinted)
+    def __init__(self, prefix: str, name: str = ""):
+        with self.add_children_as_readables():
+            self.beamsize_set_microns = epics_signal_rw(float, prefix + "VERT_REQ")
+            self.predicted_vertical_num_lenses = epics_signal_rw(
+                float, prefix + "LENS_PRED"
+            )
+            self.number_filters_sp = epics_signal_rw(int, prefix + "NUM_FILTERS")
+            self.start = epics_signal_rw(int, prefix + "START.PROC")
+            self.start_rbv = epics_signal_r(int, prefix + "START_RBV")
+            self.vertical_lens_rbv = epics_signal_r(float, prefix + "VER")
 
-    TIMEOUT = 120
-    _POLLING_WAIT = 0.01
+        self.TIMEOUT = 120
 
-    def polling_wait_on_start_rbv(self, for_value):
-        # For some reason couldn't get monitors working on START_RBV
-        # (See https://github.com/DiamondLightSource/dodal/issues/152)
-        start_time = time()
-        while time() < start_time + self.TIMEOUT:
-            RBV_value = self.start_rbv.get()
-            if RBV_value == for_value:
-                return
-            sleep(self._POLLING_WAIT)
+        super().__init__(name=name)
 
-        # last try
-        if self.start_rbv.get() != for_value:
-            raise TimeoutError()
+    async def _observe_beamsize_microns(self):
+        is_set_filters_done = False
 
-    def set(self, beamsize_microns: float) -> DeviceStatus:
+        async def set_based_on_prediction(value: float):
+            if not math.isclose(
+                self.latest_pred_vertical_num_lenses, value, abs_tol=1e-8
+            ):
+                # We can only put an integer number of lenses in the beam but the
+                # calculation in the IOC returns the theoretical float number of lenses
+                nonlocal is_set_filters_done
```

Comment:
Could: I always think `nonlocal` is a bit of a code smell. For example, in this case we could just return the value but I'm also not 100% sure of the purpose of `is_set_filters_done`